File: src/ens/lang/tokenizer/TokenType.java
package ens.lang.tokenizer;

public enum TokenType {
    EOF(),
    ERROR,
    L_PAREN("("),
    R_PAREN(")"),
    L_BRACE("{"),
    R_BRACE("}"),
    L_BRACKET("["),
    R_BRACKET("]"),
    SEMI(";"),
    COMMA(","),
    DOT("."),
    ELLIPSIS("..."),
    PLUS("+"),
    SUB("-"),
    STAR("*"),
    SLASH("/"),
    EQ("="),
    GT(">"),
    LT("<"),
    NOT("!"),
    QUES("?"),
    COLON(":"),
    EQ_EQ("=="),
    LT_EQ("<="),
    GT_EQ(">="),
    NOT_EQ("!="),
    ARROW("=>"),
    AND("&&"),
    OR("||"),
    BIT_AND("&"),
    BIT_OR("|"),
    PLUS_PLUS("++"),
    SUB_SUB("--"),
    CARET("^"),
    PERCENT("%"),
    LT_LT("<<"),
    GT_GT(">>"),
    GT_GT_GT(">>>"),
    PLUS_EQ("+="),
    SUB_EQ("-="),
    STAR_EQ("*="),
    SLASH_EQ("/="),
    BIT_AND_EQ("&="),
    BIT_OR_EQ("|="),
    CARET_EQ("^="),
    PERCENT_EQ("%="),
    LT_LT_EQ("<<="),
    GT_GT_EQ(">>="),
    GT_GT_GT_EQ(">>>="),
    BYTE("byte"),
    CHAR("char"),
    INT("int"),
    LONG("long"),
    FLOAT("float"),
    DOUBLE("double"),
    SHORT("short"),
    BOOLEAN("boolean"),
    CHAR_LITERAL,
    INT_LITERAL,
    LONG_LITERAL,
    FLOAT_LITERAL,
    DOUBLE_LITERAL,
    STRING_LITERAL,
    IDENTIFIER(),
    ASSERT("assert"),
    BREAK("break"),
    CASE("case"),
    CATCH("catch"),
    CLASS("class"),
    CONST("const"),
    CONTINUE("continue"),
    DEFAULT("default"),
    DO("do"),
    ELSE("else"),
    ENUM("enum"),
    EXTENDS("extends"),
    FINAL("final"),
    FINALLY("finally"),
    FOR("for"),
    GOTO("goto"),
    IF("if"),
    IMPORT("import"),
    NEW("new"),
    PACKAGE("package"),
    PUBLIC("public"),
    RETURN("return"),
    STATIC("static"),
    SUPER("super"),
    SWITCH("switch"),
    THIS("this"),
    THROW("throw"),
    THROWS("throws"),
    TRY("try"),
    VOID("void"),
    WHILE("while");

    private String name;

    TokenType() {
    }

    TokenType(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public static TokenType lookup(String name) {
        for (TokenType t : TokenType.values()) {
            if (name.equals(t.name)) {
                return t;
            }
        }
        return IDENTIFIER;
    }
}

File: src/tokenizer/TokenType.cpp
#include "TokenType.h"

#include <array>
#include <utility>

static constexpr std::pair<TokenType, std::u16string_view> TOKEN_NAMES[] = {
    {TokenType::L_PAREN,      u"("},
    {TokenType::R_PAREN,      u")"},
    {TokenType::L_BRACE,      u"{"},
    {TokenType::R_BRACE,      u"}"},
    {TokenType::L_BRACKET,    u"["},
    {TokenType::R_BRACKET,    u"]"},
    {TokenType::SEMI,         u";"},
    {TokenType::COMMA,        u","},
    {TokenType::DOT,          u"."},
    {TokenType::ELLIPSIS,     u"..."},
    {TokenType::PLUS,         u"+"},
    {TokenType::SUB,          u"-"},
    {TokenType::STAR,         u"*"},
    {TokenType::SLASH,        u"/"},
    {TokenType::EQ,           u"="},
    {TokenType::GT,           u">"},
    {TokenType::LT,           u"<"},
    {TokenType::NOT,          u"!"},
    {TokenType::QUES,         u"?"},
    {TokenType::COLON,        u":"},
    {TokenType::EQ_EQ,        u"=="},
    {TokenType::LT_EQ,        u"<="},
    {TokenType::GT_EQ,        u">="},
    {TokenType::NOT_EQ,       u"!="},
    {TokenType::ARROW,        u"=>"},
    {TokenType::AND,          u"&&"},
    {TokenType::OR,           u"||"},
    {TokenType::BIT_AND,      u"&"},
    {TokenType::BIT_OR,       u"|"},
    {TokenType::PLUS_PLUS,    u"++"},
    {TokenType::SUB_SUB,      u"--"},
    {TokenType::CARET,        u"^"},
    {TokenType::PERCENT,      u"%"},
    {TokenType::LT_LT,        u"<<"},
    {TokenType::GT_GT,        u">>"},
    {TokenType::GT_GT_GT,     u">>>"},
    {TokenType::PLUS_EQ,      u"+="},
    {TokenType::SUB_EQ,       u"-="},
    {TokenType::STAR_EQ,      u"*="},
    {TokenType::SLASH_EQ,     u"/="},
    {TokenType::BIT_AND_EQ,   u"&="},
    {TokenType::BIT_OR_EQ,    u"|="},
    {TokenType::CARET_EQ,     u"^="},
    {TokenType::PERCENT_EQ,   u"%="},
    {TokenType::LT_LT_EQ,    u"<<="},
    {TokenType::GT_GT_EQ,    u">>="},
    {TokenType::GT_GT_GT_EQ, u">>>="},
    {TokenType::BYTE,         u"byte"},
    {TokenType::CHAR,         u"char"},
    {TokenType::INT,          u"int"},
    {TokenType::LONG,         u"long"},
    {TokenType::FLOAT,        u"float"},
    {TokenType::DOUBLE,       u"double"},
    {TokenType::SHORT,        u"short"},
    {TokenType::BOOLEAN,      u"boolean"},
    {TokenType::ASSERT,       u"assert"},
    {TokenType::BREAK,        u"break"},
    {TokenType::CASE,         u"case"},
    {TokenType::CATCH,        u"catch"},
    {TokenType::CLASS,        u"class"},
    {TokenType::CONST,        u"const"},
    {TokenType::CONTINUE,     u"continue"},
    {TokenType::DEFAULT,      u"default"},
    {TokenType::DO,           u"do"},
    {TokenType::ELSE,         u"else"},
    {TokenType::ENUM,         u"enum"},
    {TokenType::EXTENDS,      u"extends"},
    {TokenType::FINAL,        u"final"},
    {TokenType::FINALLY,      u"finally"},
    {TokenType::FOR,          u"for"},
    {TokenType::GOTO,         u"goto"},
    {TokenType::IF,           u"if"},
    {TokenType::IMPORT,       u"import"},
    {TokenType::NEW,          u"new"},
    {TokenType::PACKAGE,      u"package"},
    {TokenType::PUBLIC,       u"public"},
    {TokenType::RETURN,       u"return"},
    {TokenType::STATIC,       u"static"},
    {TokenType::SUPER,        u"super"},
    {TokenType::SWITCH,       u"switch"},
    {TokenType::THIS,         u"this"},
    {TokenType::THROW,        u"throw"},
    {TokenType::THROWS,       u"throws"},
    {TokenType::TRY,          u"try"},
    {TokenType::VOID,         u"void"},
    {TokenType::WHILE,        u"while"},
};

std::u16string_view getTokenName(TokenType type) {
    for (const auto& [t, name] : TOKEN_NAMES) {
        if (t == type) return name;
    }
    return {};
}

TokenType lookupToken(std::u16string_view name) {
    for (const auto& [type, n] : TOKEN_NAMES) {
        if (n == name) return type;
    }
    return TokenType::IDENTIFIER;
}
